consensus/src/pos/target.rs

```diff
@@ -0,0 +1,477 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use chainstate_types::{BlockIndexHandle, BlockIndexHistoryIterator};
+use common::{
+    chain::{
+        block::{BlockHeader, ConsensusData},
+        Block, ChainConfig, GenBlockId, PoSChainConfig, PoSStatus, RequiredConsensus,
+    },
+    primitives::{BlockHeight, Compact, Id, Idable},
+    Uint256,
+};
+use itertools::Itertools;
+
+use crate::pos::error::ConsensusPoSError;
+
+fn calculate_average_block_time(
+    chain_config: &ChainConfig,
+    pos_config: &PoSChainConfig,
+    block_id: Id<Block>,
+    block_height: BlockHeight,
+    block_index_handle: &impl BlockIndexHandle,
+) -> Result<u64, ConsensusPoSError> {
+    let history_iter = BlockIndexHistoryIterator::new(block_id.into(), block_index_handle);
+
+    // Determine net upgrade version that current block belongs to.
+    // Across versions config parameters might change so it's invalid to mix versions for target calculations
+    let (_, net_version) = chain_config
+        .net_upgrade()
+        .version_at_height(block_height)
+        .expect("NetUpgrade must've been initialized");
+    let net_version_range = chain_config
+        .net_upgrade()
+        .height_range(net_version)
+        .expect("NetUpgrade must've been initialized");
+
+    // Get timestamps from the history but make sure they belong to the same consensus version
+    let block_times = history_iter
+        .take(pos_config.blocks_count_to_average())
+        .filter(|block_index| net_version_range.contains(&block_index.block_height()))
+        .map(|block_index| block_index.block_timestamp().as_int_seconds())
+        .collect::<Vec<_>>();
+
+    // At least 2 items are required to calculate difference
+    debug_assert!(block_times.len() >= 2);
+
+    // Block times are taken from history so they are sorted backwards
+    let block_diffs = block_times
+        .iter()
+        .tuple_windows::<(&u64, &u64)>()
+        .map(|t| t.0 - t.1)
+        .collect::<Vec<_>>();
```

Comment:
Same here. No need to collect.

Reply:
When going through this, please ensure you write tests for all possible invalid sizes.

Reply:
I removed one collect out of 2. Not sure how to achieve it in a clean way with iterators.